fix: Add unsubscribe functionality to `NewsletterModal`

aryanprince opened the pull request:
# Summary

This PR fixes a bug where unsubscribing from a newsletter via the newsletter modal failed to work as expected. Specifically, the bug prevented the decrement of unsubscribe credits, updating of the newsletter status, triggering of analytics and more. You can find more details about the issue [here](https://github.com/elie222/inbox-zero/issues/209).

The main issue was that the unsubscribe buttons in the app were performing different actions. With this fix, both unsubscribe buttons  (table row action and the one in the newsletter modal) now use the same `useUnsubscribeButton` hook, ensuring consistent behavior across the app.

Fixes #209

## Changes

- [x] Updated the `onClick` property for the unsubscribe button within the newsletter modal to use the `onUnsubscribe` function from the `useUnsubscribeButton` hook, ensuring all expected actions are executed.
- [x] Passed the `mutate` prop from the parent `BulkUnsubscribeSection` component to the `NewsletterModal` component.
- [x] Refactored all instances of the `NewsletterModal` component cus changes in its props, which are now required (was optional before).

## Demo

### Before the fix

The GIF below shows the unsubscribe process using the table row action button, which works as expected:

![Error - Unsubbing using direct unsub button](https://github.com/user-attachments/assets/11cc2aba-37b6-4ba5-a7fe-449ce12bc980)

In contrast, the GIF below demonstrates unsubscribing via the newsletter modal, which does not work as expected. It doesn't update the unsubscribe count or update the newsletter's status:

![Error - Unsubbing using email dialog menu](https://github.com/user-attachments/assets/365303c1-d537-415b-a2fb-b0577d75ba85)

### After the fix

The GIF below illustrates the corrected behavior after applying this PR. Unsubscribing via the newsletter modal now does all of the same actions as clicking the unsubscribe button from the table row action button:

![Fixed - Unsubbing using email dialog menu](https://github.com/user-attachments/assets/95d3d46b-e1c0-48d8-b3cd-be9060019710)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced conditional rendering for the `NewsletterModal` across multiple components to enhance performance and user experience.
	- Added a required `newsletter` prop and an optional `mutate` prop to the `NewsletterModal`, improving data handling and interactivity.
	- Enhanced the unsubscribe button with a dynamic loading state for better user feedback during the unsubscribe process.
	- Integrated new hooks for managing subscription access and the unsubscribe process in the `NewsletterModal`.

- **Bug Fixes**
	- Improved clarity and efficiency by eliminating unnecessary renders of the `NewsletterModal` when conditions are not met.
  
- **Style**
	- Minor formatting adjustments in the `ColorSelector` and `LinkSelector` components for improved consistency.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/elie222/inbox-zero?pullRequest=210) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/elie222/inbox-zero?pullRequest=210) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/elie222/inbox-zero?pullRequest=210) it.</sub>

aryanprince commented on apps/web/app/(app)/stats/NewsletterModal.tsx:
```diff
@@ -21,27 +21,46 @@ import { SectionHeader } from "@/components/Typography";
+  newsletter: Pick<Row, "name" | "lastUnsubscribeLink" | "autoArchived">;
   onClose: (isOpen: boolean) => void;
   refreshInterval?: number;
+  mutate?: () => Promise<any>;
```
I agree bot. I tried my best to make the suggestion of `mutate?: () => Promise<void>;` work but there were a lot of side effects with it. Would be happy to see a fix for this.

The issue here is that the `mutate` function passed down to this `NewsletterModal` component is to mutate multiple newsletters, but this component only needs to mutate a single newsletter.

The only fixes I could think of was to: 

1. Create a new API route to mutate a single newsletter as well, along with the existing API route to mutate multiple newsletters.
2. Update the `useUnsubscribeButton` hook's logic, and all of its usages.

I think this will need refactoring in the future regardless. Would be happy to have a look at this later.

elie222 commented on apps/web/app/(app)/stats/NewsletterModal.tsx:
```diff
@@ -57,8 +76,15 @@ export function NewsletterModal(props: {
+                  onClick={onUnsubscribe}
                 >
-                  Unsubscribe
+                  {!unsubscribeLoading && <span>Unsubscribe</span>}
+                  {!!unsubscribeLoading && (
+                    <div className="flex cursor-not-allowed items-center opacity-50">
+                      <ButtonLoader />
+                      <span>Unsubscribing...</span>
+                    </div>
+                  )}
```
Could use a ternary to make this cleaner. ie: `unsubscribeLoading ? Unsub : Loading...`

If you're trying to add disabled state, then you could add disabled={unsubscribeLoading} to the button itself.

aryanprince replied:
Updated it now! Used a ternary here instead and added the disabled state to the button.

elie222 commented on apps/web/app/(app)/bulk-unsubscribe/BulkUnsubscribeSection.tsx:
```diff
@@ -223,11 +223,14 @@ export function BulkUnsubscribeSection(props: {
-        onClose={() => setOpenedNewsletter(undefined)}
-        refreshInterval={props.refreshInterval}
-      />
+      {openedNewsletter && (
```
Did you need to add `{openedNewsletter && (`?
I think it stops the animation when it's done like that. Or was this added to fix a bug?

aryanprince replied:
# Reason

This is cus the `<NewsletterModal />` component now requires (non-optional) the `openedNewsletter` prop, which was previously optional. 

The reason for this change is that the component now uses the existing `useUnsubscribeButton` hook, which **does need a required** `openedNewsletter` prop.

Initially, I tried keeping openedNewsletter as optional (to have less diffs like the one you brought up here), but that would have required introducing a second hook for unsubscribing, which would complicate the logic unnecessarily. 

## Minor Benefit

Additionally, making `openedNewsletter` a required prop also ensures that the `<NewsletterModal />` is **only** rendered when all necessary data is present, making it more robust and predictable in terms of when it renders throughout the app.

## Quick Demo

Also, below GIF shows that the animations are still intact. No issues with that according to my testing.

![CleanShot 2024-08-23 at 12 29 31](https://github.com/user-attachments/assets/a51ee935-77a1-4f9f-ac81-649a91360b20)

elie222 reviewed (changes requested):
2 comments

aryanprince commented on apps/web/app/(app)/stats/NewsletterModal.tsx:
```diff
@@ -1,47 +1,66 @@
-  DialogHeader,
-  DialogTitle,
-} from "@/components/ui/dialog";
+  MoreDropdown,
+  useUnsubscribeButton,
+} from "@/app/(app)/bulk-unsubscribe/common";
```
Bot, this is incorrect. `MoreDropdown` is a React component and `useUnsubscribeButton` is a custom React hook.

elie222 commented on apps/web/app/(app)/stats/NewsletterModal.tsx:
```diff
@@ -52,13 +71,23 @@ export function NewsletterModal(props: {
                 >
-                  Unsubscribe
+                  {unsubscribeLoading ? (
+                    // Button loading state
```
really don't need this comment lol. pretty clear from the code that it's loading state 😆

aryanprince replied:
Haha done, removed the comments now

elie222 commented on apps/web/app/(app)/stats/NewsletterModal.tsx:
```diff
@@ -52,13 +71,23 @@ export function NewsletterModal(props: {
+                      <span>Unsubscribing...</span>
+                    </div>
+                  ) : (
+                    // Button default state
```
yes lol. you can remove this comment

aryanprince replied:
Removed this comment too

elie222 commented:
Hey, took a look again. I'm going to run myself locally and then will get it merged in after. I see there's also merge conflicts now :(

aryanprince commented:
> [!NOTE] 
> Just a quick update for this PR

# Summary

✅ Resolved conflicts, updated changes in this PR according to new code changes from main.

## Major Changes

- [x] In terms of changes from my last commit on this PR (merge conflicts):
	- [x] 	The old `useUnsubscribeButton` hook is now replaced by a newer `useUnsubscribe` hook instead.
	- [x] 	Previously there was a `<ButtonLoader2 />` component for a loading state, now it's just a prop on the `<Button loading={true} />` component.

No other changes to this PR other than these.

## Minor Changes

There were only 5 file changes in this PR previously, now there's 7. 

2 of the extra files changed are due to `lint-staged` step which forces me to format code in files `apps/web/app/(app)/compose/selectors/color-selector.tsx` and `apps/web/app/(app)/compose/selectors/link-selector.tsx`.

aryanprince commented on apps/web/app/(app)/cold-email-blocker/ColdEmailList.tsx:
```diff
@@ -144,12 +144,12 @@ export function ColdEmailList() {
-            }
-            onClose={() => setOpenedRow(undefined)}
-          />
+          {openedRow && (
+            <NewsletterModal
+              newsletter={openedRow && { name: openedRow.fromEmail || "" }}
+              onClose={() => setOpenedRow(undefined)}
+            />
+          )}
```
Good suggestion, resolved!

aryanprince commented on apps/web/app/(app)/new-senders/NewSenders.tsx:
```diff
@@ -211,11 +211,13 @@ export function NewSenders({ refreshInterval }: { refreshInterval: number }) {
+          onClose={() => setOpenedNewsletter(undefined)}
+          refreshInterval={props.refreshInterval}
+        />
+      )}
```
Thanks, fixed it now!

elie222 commented:
Closing as it's very old with a lot of conflicts.
